src/Clause.ts

```diff
@@ -300,11 +342,26 @@ export default class Clause extends Builder {
           message: `duplicate definition ${JSON.stringify(clause.aoid)}`,
         });
       } else {
+        const signature = clause.signature;
         const op: PartialBiblioEntry = {
           type: 'op',
           aoid: clause.aoid,
           refId: clause.id,
+          signature,
+          _node: clause.node,
         };
+        if (
+          signature?.return?.kind === 'union' &&
+          signature.return.types.some(e => e.kind === 'completion') &&
+          signature.return.types.some(e => e.kind !== 'completion')
+        ) {
+          spec.warn({
+            type: 'node',
+            node: clause.header!,
+            ruleId: 'completion-union',
+            message: `algorithms should return either completions or things which are not completions, never both`,
```

Comment:
I think this warning would be unnecessary if we can get the type system to not even allow construction of this kind of signature.